### DISCUSSION TITLE
Convert State Machine to `statig`?

### DISCUSSION THREAD

**Discussion by akennedy4155** [ArroyoSystems/arroyo#58] [General]
I've used `statig`, which is [open source](https://github.com/mdeloof/statig), to implement FSM in the past with Rust, and I loved the user experience.  I noticed that the state/transition logic in the controller is not using any library as its basis.  It's all code written for Arroyo.

Would it make more sense to use an external library for such logic, so that we keep Arroyo specific for streaming and let the FSM logic be maintained by a project that specifically focuses on such?

**Comment by mwylde** [comment]
The state machine implementation is based on the approach from this blog post: https://deislabs.io/posts/a-fistful-of-states.

It has a number of nice properties, in particular it allows compile-time checking of valid state transitions and allows actions on transitions.

The state machine implementation serves our needs, and is only ~50 lines of code, so I'm not sure it's worth adopting some other library for that.

**Reply by akennedy4155** [reply]
Haven't seen that article before.  I'll have to go check that out!  I'm fine if you want to lock this with that response, sounds like a case of if it ain't broke don't fix it :D
